js/mod.ts

```diff
@@ -217,3 +245,57 @@ function checkCompilerOptions(
     );
   }
 }
+
+interface SerializedImportMap {
+  url: string;
+  value: string;
+}
+
+async function fetchImportMap(
+  urlOrString: URL | string | undefined,
+): Promise<SerializedImportMap | undefined> {
+  if (!urlOrString) return undefined;
+  const url = urlOrString instanceof URL
+    ? urlOrString
+    : toFileUrl(resolve(urlOrString));
+  const fetchImportMapResponse = await fetch(url);
+  const body = await fetchImportMapResponse.json();
+  return {
+    url: url.toString(),
+    value: JSON.stringify({
+      imports: body.imports,
+      scopes: body.scopes,
+    }),
+  };
+}
+
+async function buildImportMap(
+  options: {
+    baseUrl?: URL | string;
+    imports?: Record<string, string>;
+    scopes?: Record<string, Record<string, string>>;
+    importMap?: URL | string;
+  },
+): Promise<SerializedImportMap | undefined> {
+  const { baseUrl, imports, scopes, importMap } = options;
+  if (imports != null || scopes != null) {
+    const url = baseUrl instanceof URL
+      ? new URL(baseUrl.toString())
+      : toFileUrl(baseUrl ? resolve(baseUrl) : Deno.cwd());
+    // Rust lib expects url to be the file URL to the import map file, but the
+    // JS API expects it to be the file URL to the root directory, so we need to
+    // append an extra slash.
+    if (!url.pathname.endsWith("/")) {
+      url.pathname += "/";
+    }
```

Comment:
I don't like this but, I am not a fan of not doing it either. Getting a non-descriptive error because of a missing slash feels like a really poor user experience.

Reply:
The rust url crate expects a trailing slash in order to be considered a directory and this will always be a directory so this is ok.